### PR TITLE
fix(approvals): persist unverified_contact when guardian denies access (LUM-2656)

### PR DESCRIPTION
**Pull request by ashleeradka**
Fixes [LUM-2656](https://linear.app/vellum/issue/LUM-2656/denying-a-contact-verification-prompt-should-persist-unverified).

## Prompt / plan

When an unknown sender (e.g. a bot) DMs the assistant, the guardian gets an access/verification prompt. If the guardian **denies**, the sender should become a persisted `unverified_contact` — known but not verified — and should **not** re-trigger the prompt on subsequent DMs.

**Root cause (confirmed):** the `access_request` deny path in `guardian-request-resolvers.ts` delivered denial notifications/lifecycle signals but **persisted nothing**. The sender stayed `unknown`, so every subsequent DM created a fresh access request and re-prompted the guardian.

The fix has two complementary parts, matching the issue's Investigation TODO ("fix it so [deny] persists `unverified_contact` … and ensure the prompt is suppressed for `unverified_contact` on inbound"):

**1. Deny persists `unverified_contact` (gateway-first).** The `accessRequestResolver` deny path now seeds the sender via a new `seedUnverifiedMemberChannel` relay (`assistant/src/contacts/member-write-relay.ts`) → gateway `create_contact` IPC → `ContactStore.upsertContact`. A brand-new channel lands at status `unverified` (→ trust class `unverified_contact`, per `resolveTrustVerdict` / `actor-trust-resolver.ts`); an existing channel's status is **preserved**, so a blocked / revoked / active row is never reactivated or downgraded. Best-effort: a failed relay is logged and swallowed so it can never fail the guardian's deny. Skipped for desktop-origin (`vellum`) requests, which carry no channel identity.

**2. Suppress re-prompting after a terminal deny.** `notifyGuardianOfAccessRequest` (`assistant/src/runtime/access-request-helper.ts`) now skips creating a new prompt when a terminally-`denied` `access_request` already exists for the same sender/channel/assistant (scoped by the same assistant-scoped `conversationId` as the existing pending-dedup). Callers already treat `notified: false` as "deliver the plain canned reply" and (voice) "fail closed", so a previously-denied sender is denied without re-notifying the guardian.

Behavior preserved:
- `unverified_contact` remains admission-only — it is treated identically to `trusted_contact` for every downstream capability/tool/approval decision (`capabilities.ts`). Under permissive admission floors (`any_contact`, `strangers`) the persisted sender is now admitted at the unverified tier; under stricter floors they are denied but no longer re-prompt.
- The guardian can still **manually** verify the contact later — that path (`mark_channel_verified` / control plane) does not run through `notifyGuardianOfAccessRequest`, and once verified the sender is `trusted_contact` (no access-request path at all).
- A deny for one sender/channel does not suppress prompts for other senders or other channels.

## Test plan

- **New unit tests** — `seedUnverifiedMemberChannel` relays to `create_contact` with `channelType`/`address`/`displayName`, omits `displayName` when absent, and swallows gateway errors (best-effort). `assistant/src/contacts/__tests__/member-write-relay.test.ts` → **9 pass**.
- **New behavioral tests** — after a prior `denied` `access_request`, `notifyGuardianOfAccessRequest` returns `{ notified: false, reason: "already_denied" }` with no new signal/pending request; a different sender still gets a fresh prompt; a denial on one channel does not suppress another channel. `assistant/src/__tests__/non-member-access-request.test.ts` → **21 pass**.
- **Regression** — ran individually (green): `trusted-contact-lifecycle-notifications` (deny path), `acl-enforcement`, `admission-policy`, `trusted-contact-multichannel`, `trusted-contact-inline-approval-integration`, `slack-inbound-verification`, `canonical-guardian-store`, `trusted-contact-verification`.
- `bunx prettier`, `eslint`, and the pre-commit type-check all pass on the changed files.

<sub>Note: 4 failures seen only when running a large batch of these files in one `bun test` invocation reproduce identically on `main` (a pre-existing cross-file `mock.module` isolation artifact); all pass when run per-file.</sub>

## CLI verb checklist

N/A — no new IPC route under `assistant/src/runtime/routes/`. The deny path reuses the existing gateway `create_contact` IPC via a daemon-side relay.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01FT3W9EuWZaRgcCaro6kuV1

---
_Generated by [Claude Code](https://claude.ai/code/session_01FT3W9EuWZaRgcCaro6kuV1)_

### CODE REVIEW

**Comment by ashleeradka** [comment]
Thanks for the reviews — addressed the automated feedback in `ec3e174`:

- **Codex (P2 — challenge fires before the deny-suppression):** correct catch. Slack/email self-verify challenges now short-circuit for a terminally-denied sender *before* a session is minted, so a denied sender no longer gets an unusable "share a 6-digit code" DM the guardian was never told about. Factored the check into a shared `isAccessRequestDenied` helper (single source of truth for the assistant-scoped `conversationId`), reused by `notifyGuardianOfAccessRequest`; blocked members skip the lookup. Added an integration test asserting a terminally-denied Slack sender gets no challenge, no session, and no guardian re-prompt.
- **Devin (voice caller gets a false "I'll let them know" promise):** correct. A previously-denied caller who dials again now routes to `handleAccessRequestDenied()` (delivers the denial copy + hangs up) instead of the timeout path, and the log no longer says "no sender ID" for the `already_denied` case.
- **Devin (temporal test comment):** rewritten in present tense per the code-comments rule.
- **Devin (optional — seed `displayName` from `actorDisplayName`):** deferred to a follow-up rather than folded in here. Threading the sender's display name through the canonical request into the seed overlaps with the contact/assistant-typing work planned next (bots/assistants currently seed as `human`), so it's cleaner to group them. Not a correctness issue, as noted.

Full green locally: added coverage plus the existing regressions pass per-file.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ashleeradka** [comment]
Follow-up on the Codex re-review of `ec3e174` (mock-completeness finding): addressed in `26720a5` — added `isAccessRequestDenied` to the `access-request-helper` mock in `trusted-contact-multichannel.test.ts`.

Note the suite wasn't actually failing: it initializes a real DB, so the unmocked `isAccessRequestDenied` fell through to the real helper and read an empty table (returning `false`) rather than crashing. Stubbing it keeps the mock complete and matches the `acl-enforcement.test.ts` mock, so the intent (fully mocked `access-request-helper`) is explicit.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: assistant/src/approvals/guardian-request-resolvers.ts [hunk 1]
```diff
@@ -499,6 +502,23 @@ const accessRequestResolver: GuardianRequestResolver = {
...
 
+      // Persist the denied sender as an unverified_contact (gateway-first).
+      // Denial is a terminal decision: the sender becomes a known, unverified
+      // contact so future inbound resolves as unverified_contact rather than
+      // re-triggering discovery. Paired with the denied-request suppression in
+      // notifyGuardianOfAccessRequest, this stops the prompt from re-firing on
+      // every subsequent DM. The guardian can still verify them later. Skipped
+      // for desktop-origin (vellum) requests, which carry no channel identity.
+      if (requesterExternalUserId && channel !== "vellum") {
+        await seedUnverifiedMemberChannel({
```
[the remark this answers is not in the document]
Comment (ashleeradka) [reply assistant/src/approvals/guardian-request-resolvers.ts:513]: Fixed in d7a38b5. The terminal-deny check now runs *ahead* of the policy-aware admission bypass in both the non-member and inactive-member paths of `enforceIngressAcl`, and the bypass is gated on `!terminallyDenied`.

So a guardian-denied sender — even though they're persisted as an `unverified_contact` (rank 2) — is no longer passed through to the admission floor: on an `any_contact` channel (floor 2) they're held out and denied at the ACL instead of clearing the floor. They stay saved and verifiable later; they just can't DM/@mention on any policy until the guardian verifies them. A *normal* (non-denied) unverified contact is still bypassed and admitted where the policy allows — only the explicitly-denied are carved out.

Added regression tests in `acl-enforcement.test.ts` covering denied-vs-non-denied under `any_contact` (inactive member) and `strangers` (non-member), asserting the denied case is denied while the non-denied case is still admitted.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: assistant/src/runtime/access-request-helper.ts [hunk 2]
```diff
@@ -147,6 +147,28 @@ export async function notifyGuardianOfAccessRequest(
...
+    sourceChannel,
+    kind: "access_request",
+    conversationId,
+  });
+  if (deniedCanonical.length > 0) {
+    log.debug(
+      { sourceChannel, actorExternalId, deniedId: deniedCanonical[0].id },
+      "Suppressing access request notification — guardian already denied this sender",
+    );
+    return { notified: false, reason: "already_denied" };
```
[the remark this answers is not in the document]
Comment (ashleeradka) [reply assistant/src/runtime/access-request-helper.ts:216]: Addressed — the terminal-deny precheck now runs ahead of challenge creation in the caller (`acl-enforcement.ts`), which is the "expose a precheck" option you suggested.

Both the non-member and inactive-member branches compute `terminallyDenied` via `isAccessRequestDenied(...)` and gate `initiateSlackVerificationChallenge` / `initiateEmailVerificationChallenge` on `!terminallyDenied`. So for a denied sender whose earlier session expired, no new verification session is created — they get only the canned rejection reply, never an unusable "share a 6-digit code" DM, and the guardian is not re-notified. (Challenge gating landed in ec3e174; d7a38b5 additionally hoists the same check ahead of the admission-floor bypass.)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: assistant/src/runtime/routes/inbound-stages/acl-enforcement.ts [hunk 3]
```diff
@@ -23,7 +23,10 @@ import { resolveGuardianName } from "../../../prompts/user-reference.js";
 import { getLogger } from "../../../util/logger.js";
 import { truncate } from "../../../util/truncate.js";
 import { hashVoiceCode } from "../../../util/voice-code.js";
-import { notifyGuardianOfAccessRequest } from "../../access-request-helper.js";
+import {
+  isAccessRequestDenied,
+  notifyGuardianOfAccessRequest,
+} from "../../access-request-helper.js";
```
[the remark this answers is not in the document]
Comment (ashleeradka) [reply assistant/src/runtime/routes/inbound-stages/acl-enforcement.ts:29]: Done in 26720a5 — `trusted-contact-multichannel.test.ts` now stubs `isAccessRequestDenied` alongside `notifyGuardianOfAccessRequest` in the `access-request-helper.js` mock, so the module loads with the new named import. Verified the scoped suite is green (11 pass).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ashleeradka** [comment]
Addressed the Codex **P1** (denied senders reaching the `any_contact` floor) in `d7a38b5`.

**Root cause:** persisting a denied sender as `unverified_contact` (rank 2) was correct for suppressing the re-prompt, but on a channel with an `any_contact` floor (rank 2) the admission stage would then admit them — the deny silently became standing access.

**Fix:** hoist the terminal-deny check ahead of the policy-aware admission bypass in `enforceIngressAcl` (both the non-member and inactive-member paths) and gate the bypass on `!terminallyDenied`. A guardian-denied sender is now held out of the admission floor and denied at the ACL, on every policy, until the guardian verifies them — while a *normal* (non-denied) unverified contact is still admitted where the policy allows. The non-member gate also covers the edge case where the `unverified_contact` seed failed and the sender resurfaces as a stranger.

Added regression tests in `acl-enforcement.test.ts`: a terminally-denied unverified member under `any_contact` is denied while a non-denied one is admitted, with the same contrast on the non-member/`strangers` path. Typecheck + affected suites green per-file; all three open review threads replied to and resolved.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: assistant/src/approvals/guardian-request-resolvers.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ashleeradka) [reply assistant/src/approvals/guardian-request-resolvers.ts:513]: Update — reversing this in `b99c30e`. After reviewing against the Slack-permissions PRD (Lane A), we're intentionally **not** gating admission on the denied request.

Per the PRD, a guardian deny persists the sender as an `unverified_contact` and admission stays pure rank-vs-floor: a denied sender is admitted on the same terms as any other `unverified_contact` (admitted under `any_contact`/`strangers`, denied under stricter floors). The deny's role is to stop **re-prompting** the guardian — not to exclude the sender from a permissive channel. Keeping a contact out of every floor is the separate **block** action's job (`block → revoked`), which isn't wired to the access-request card yet but is the PRD's designated exclusion mechanism.

So `b99c30e` reverts the admission-floor gate from d7a38b5 (both the non-member and inactive-member bypasses are ungated again). The re-prompt suppression stays: `notifyGuardianOfAccessRequest` still returns `already_denied`, and the Slack/email self-verify challenge is still skipped for a denied sender in the deny path — a denied sender under a strict floor gets only the canned reply, never an unusable challenge.

Unresolving so reviewers can confirm the direction: this deliberately declines the P1 in favor of the PRD's deny-vs-block split.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (ashleeradka) [reply assistant/src/approvals/guardian-request-resolvers.ts:513]: Resolving — settled by design. Confirmed against the PRD: `deny → unverified_contact`, and admission stays pure rank-vs-floor, so a denied sender is admitted wherever `unverified_contact` clears the floor (e.g. `any_contact`). `block → revoked` is the mechanism for keeping someone out everywhere — deny is not a soft-block. `b99c30e` reverted the admission-floor gate accordingly (re-prompt + Slack/email challenge suppression stayed).

The full introduction-card design that makes this coherent (Trust / Verify-with-a-code / Leave-unverified / Block, with a signal-driven handshake) is tracked in LUM-2670. Closing this as intended-behavior, not a bug.

---
_Generated by [Claude Code](https://claude.ai/code)_
